src/library/export/libraryexporter.cpp

```diff
@@ -0,0 +1,59 @@
+#ifdef __DJINTEROP__
+#include "library/export/libraryexporter.h"
+
+#include <QProgressDialog>
+#include <QThreadPool>
+
+#include "library/export/enginelibraryexportjob.h"
+#include "library/trackloader.h"
+
+namespace mixxx {
+
+LibraryExporter::LibraryExporter(QWidget* parent,
+        UserSettingsPointer pConfig,
+        TrackCollectionManager& trackCollectionManager)
+        : QWidget{parent},
+          m_pConfig{std::move(pConfig)},
+          m_trackCollectionManager{trackCollectionManager},
+          m_pTrackLoader{nullptr} {
+    m_pTrackLoader = new TrackLoader(&m_trackCollectionManager, this);
+}
+
+void LibraryExporter::requestExport() {
+    if (!m_pDialog) {
+        m_pDialog = make_parented<DlgLibraryExport>(this, m_pConfig, m_trackCollectionManager);
+        connect(m_pDialog.get(),
+                SIGNAL(startEngineLibraryExport(EngineLibraryExportRequest)),
+                this,
+                SLOT(beginEngineLibraryExport(EngineLibraryExportRequest)));
+    } else {
+        m_pDialog->show();
+        m_pDialog->raise();
+        m_pDialog->setWindowState(
+                (m_pDialog->windowState() & ~Qt::WindowMinimized) |
+                Qt::WindowActive);
+    }
+}
+
+void LibraryExporter::beginEngineLibraryExport(
+        EngineLibraryExportRequest request) {
+    // Note that the job will run in a background thread.
+    auto* pJobThread = new EngineLibraryExportJob{
+            this, m_trackCollectionManager, *m_pTrackLoader, std::move(request)};
+    connect(pJobThread, &EngineLibraryExportJob::finished, pJobThread, &QObject::deleteLater);
+
+    // Construct a modal dialog to monitor job progress.
+    // TODO(mr-smidge) - dialog doesn't appear to update with new progress until after track export?
+    auto *pd = new QProgressDialog(this);
+    pd->setLabelText(tr("Exporting to Engine Library..."));
+    pd->setMinimumDuration(0);
+    connect(pJobThread, &EngineLibraryExportJob::jobMaximum, pd, &QProgressDialog::setMaximum);
+    connect(pJobThread, &EngineLibraryExportJob::jobProgress, pd, &QProgressDialog::setValue);
+    connect(pJobThread, &EngineLibraryExportJob::finished, pd, &QObject::deleteLater);
+    connect(pd, &QProgressDialog::canceled, pJobThread, &EngineLibraryExportJob::cancel);
```

Comment:
This may be due to my inexperience with QT, but this modeless progress dialog does not actually visibly update for me.  The dialog appears, but remains undrawn until the job completes.

Any suggestions on how to diagnose/resolve this are greatly appreciated.

Reply:
QProgressDialog::setValue is a function, not a slot. I don't know if this is the culprit, but you can try to wrap it in a lambda function for testing.

Reply:
Both `setMaximum` and `setValue` are [definitely slots](https://doc.qt.io/qt-5/qprogressdialog.html#public-slots), but after some quick printf-debugging I observe that all the progress update slot invocations happen _after_ the export has completed.

I have also noticed that every invocation of `EngineLibraryExportJob::trackLoaded()` appears to be happening on the main thread!  Not sure why.  So I will investigate this first...

Reply:
This issue is now fixed, and the progress dialog updates correctly.

The track export was unintentionally happening on the GUI thread rather than the worker thread, which was entirely down to my inexperience with QThread.  The `trackLoaded()` slot now still executes on the main thread, but merely adds the loaded track to a queue for subsequent processing on the worker thread.

I also implemented cancellation.